rclpy/src/rclpy/clock.cpp

```diff
@@ -0,0 +1,244 @@
+  if (!clock) {
+    // Warning should use line number of the current stack frame
+    int stack_level = 1;
+    PyErr_WarnFormat(
+      PyExc_RuntimeWarning, stack_level, "Failed to get clock pointer in destructor");
+    rcl_reset_error();
+    return;
+  }
+
+  rcl_ret_t ret_clock = rcl_clock_fini(clock);
+  PyMem_Free(clock);
+  if (ret_clock != RCL_RET_OK) {
+    // Warning should use line number of the current stack frame
+    int stack_level = 1;
+    PyErr_WarnFormat(
+      PyExc_RuntimeWarning, stack_level, "Failed to fini clock: %s",
+      rcl_get_error_string().str);
+    rcl_reset_error();
+  }
+}
+
+py::capsule
+create_clock(rcl_clock_type_t clock_type)
+{
+  auto deleter = [](rcl_clock_t * ptr) {_rclpy_destroy_clock(ptr);};
+  auto clock = std::unique_ptr<rcl_clock_t, decltype(deleter)>(
+    static_cast<rcl_clock_t *>(PyMem_Malloc(sizeof(rcl_clock_t))),
+    deleter);
+  if (!clock) {
+    throw std::bad_alloc();
+  }
+
+  rcl_allocator_t allocator = rcl_get_default_allocator();
+  rcl_ret_t ret = rcl_clock_init(clock_type, clock.get(), &allocator);
+  if (ret != RCL_RET_OK) {
+    throw RCLError("failed to initialize clock");
+  }
+
+  PyObject * pyclock_c =
+    rclpy_create_handle_capsule(clock.get(), "rcl_clock_t", _rclpy_destroy_clock);
+  auto pyclock = py::reinterpret_steal<py::capsule>(pyclock_c);
+  // pyclock now owns the rcl_clock_t
+  clock.release();
+  return pyclock;
+}
+
+py::capsule
+clock_get_now(py::capsule pyclock)
+{
+  auto clock = static_cast<rcl_clock_t *>(
+    rclpy_handle_get_pointer_from_capsule(pyclock.ptr(), "rcl_clock_t"));
+  if (!clock) {
+    throw py::error_already_set();
+  }
+
+  int64_t nanoseconds;
+  rcl_ret_t ret = rcl_clock_get_now(clock, &nanoseconds);
+  if (ret != RCL_RET_OK) {
+    throw RCLError("failed to get current clock value");
+  }
+
+  return create_time_point(nanoseconds, clock->type);
+}
+
+bool
+clock_get_ros_time_override_is_enabled(py::capsule pyclock)
+{
+  auto clock = static_cast<rcl_clock_t *>(
+    rclpy_handle_get_pointer_from_capsule(pyclock.ptr(), "rcl_clock_t"));
+  if (!clock) {
+    throw py::error_already_set();
+  }
+
+  bool is_enabled;
+  rcl_ret_t ret = rcl_is_enabled_ros_time_override(clock, &is_enabled);
+  if (ret != RCL_RET_OK) {
+    throw RCLError("failed to get ROS time override status");
+  }
+
+  return is_enabled;
+}
+
+void
+clock_set_ros_time_override_is_enabled(py::capsule pyclock, bool enabled)
+{
+  auto clock = static_cast<rcl_clock_t *>(
+    rclpy_handle_get_pointer_from_capsule(pyclock.ptr(), "rcl_clock_t"));
+  if (!clock) {
+    throw py::error_already_set();
+  }
+
+  rcl_ret_t ret;
+  if (enabled) {
+    ret = rcl_enable_ros_time_override(clock);
+  } else {
+    ret = rcl_disable_ros_time_override(clock);
+  }
+  if (ret != RCL_RET_OK) {
+    throw RCLError("failed to set ROS time override");
+  }
+  if (PyErr_Occurred()) {
+    // Time jump callbacks raised
+    throw py::error_already_set();
+  }
```

Comment:
Could another unrelated failure cause `PyErr_Occurred` to be true? Is this something we could tackle in the class rewrite?